### PR TITLE
BIP 34: Check last minimal encoding accoriding of heights

### PR DESCRIPTION
**Pull request by sanket1729**
See: https://github.com/bitcoin/bips/blob/master/bip-0034.mediawiki

The BIP was updated with commit: https://github.com/bitcoin/bips/commit/08844fd6eff55e756c453c6d2e16de40e8f80e73

Found while reviewing #1240 

Also cleanly deals with negative heights as errors.

### CODE REVIEW

File: src/blockdata/block.rs [hunk 1]
```diff
@@ -332,6 +337,12 @@ impl Block {
         let push = input.script_sig.instructions_minimal().next().ok_or(Bip34Error::NotPresent)?;
         match push.map_err(|_| Bip34Error::NotPresent)? {
             script::Instruction::PushBytes(b) if b.len() <= 8 => {
+                // Check that the number is encoded in the minimal way.
+                // As the number is encoded in little-endian, the last
+                // byte is the most significant one.
+                if b.last() == Some(&0x00) {
+                    return Err(Bip34Error::NonMinimalPush(b.to_vec()));
+                }
```
Comment (apoelstra) [inline src/blockdata/block.rs:345]: Is this correct? Can't a number start with `0080` to force it to be positive if it'd otherwise be negative?

Comment (sanket1729) [reply src/blockdata/block.rs:345]: Good catch! This is incorrect.

File: src/blockdata/block.rs [hunk 2]
```diff
@@ -313,11 +313,16 @@ impl Block {
     }
 
     /// Returns the block height, as encoded in the coinbase transaction according to BIP34.
+    //
+    // Note:
```
Comment (dpc) [inline src/blockdata/block.rs:317]: Any reason not to return an error instead?

**Comment by dpc** [comment]
I'm confused if "accoriding of heights" is just a mistake, or there's some meaning there that I'm missing.

File: src/blockdata/block.rs (hunk 1 above)
Comment (sanket1729) [reply src/blockdata/block.rs:346]: Fixed this in a much cleaner way

**Comment by sanket1729** [comment]
Sorry about the poor description. Fixed this in two steps:
1) Make sure that decoding `CScriptNums` is always done in a minimal way. 
2) BIP34 logic to use the CScriptNum decoding to read it.

File: src/blockdata/block.rs [hunk 3]
```diff
@@ -331,14 +331,14 @@ impl Block {
...
-                // Expand the push to exactly 8 bytes (LE).
-                let mut full = [0; 8];
-                full[0..b.len()].copy_from_slice(b);
-                Ok(util::endian::slice_to_u64_le(&full))
-            }
-            script::Instruction::PushBytes(b) if b.len() > 8 => {
-                Err(Bip34Error::UnexpectedPush(b.to_vec()))
+            script::Instruction::PushBytes(b) => {
+                // Check that the number is encoded in the minimal way.
+                let h = script::read_scriptint(b).map_err(|_e| Bip34Error::UnexpectedPush(b.to_vec()))?;
```
Comment (sanket1729) [inline src/blockdata/block.rs:336]: I am unsure if I want to propagate the script error here. It has many other invariants that are unreadable from this particular function.

Comment (apoelstra) [reply src/blockdata/block.rs:336]: I think what you've done is probably the right thing.

**Review by apoelstra — APPROVED** [review]
ACK 840520042a84d38fa8097faf47a9d562004d1e7b

File: src/blockdata/script.rs [hunk 4]
```diff
@@ -225,9 +225,9 @@ impl From<bitcoinconsensus::Error> for Error {
     }
 }
 
-/// Helper to encode an integer in script format.
+/// Helper to encode an integer in script(minimal CScriptNum) format.
```
Comment (tcharding) [inline src/blockdata/script.rs:228]: Can I ask a question, why is the open parenthesis used with no space before it in docs? I've seen this elsewhere and wondered why its done like that, this is the first PR I've seen add one - hence my question. Cheers

**Review by tcharding — APPROVED** [review]
ACK 840520042a84d38fa8097faf47a9d562004d1e7b

File: src/blockdata/script.rs (hunk 4 above)
Comment (sanket1729) [reply src/blockdata/script.rs:228]: I had no idea that we are supposed to put spaces there. Won't be dismissing the two ACKs for this nit

Comment (apoelstra) [reply src/blockdata/script.rs:228]: There should be a space there -- but agreed, we can fix this in a later doc-fixup PR.

Comment (tcharding) [reply src/blockdata/script.rs:228]: No need to fix it, I can do them when I'm cleaning up docs. I just wanted to make sure it wasn't for a reason I did not know about. Cheers

**Pull request merged**